subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -82,35 +91,61 @@ protected boolean handleVirtualFileSystemContentsChanged(Collection<FileSystemLo
     }
 
     @Override
-    protected SnapshotHierarchy doUpdateVfsOnBuildStarted(SnapshotHierarchy root) {
-        return movedHierarchyHandler.handleMovedHierarchies(root);
+    public SnapshotHierarchy updateVfsOnBuildFinished(SnapshotHierarchy root, WatchMode watchMode, int maximumNumberOfWatchedHierarchies) {
+        SnapshotHierarchy newRoot = super.updateVfsOnBuildFinished(root, watchMode, maximumNumberOfWatchedHierarchies);
+        LOGGER.info("Watched directory hierarchies: {}", watchedHierarchies);
+        return newRoot;
     }
 
     @Override
-    protected void startWatchingHierarchies(Collection<File> hierarchiesToStartWatching) {
-        fileWatcher.startWatching(hierarchiesToStartWatching);
-    }
+    protected FileHierarchySet updateWatchedHierarchies(SnapshotHierarchy root) {
+        FileHierarchySet newWatchedFiles = super.updateWatchedHierarchies(root);
 
-    @Override
-    protected void stopWatchingHierarchies(Collection<File> hierarchiesToStopWatching) {
-        if (!fileWatcher.stopWatching(hierarchiesToStopWatching)) {
-            LOGGER.debug("Couldn't stop watching directories: {}", hierarchiesToStopWatching);
+        ImmutableSet<File> oldWatchedHierarchies = watchedHierarchies;
+        ImmutableSet.Builder<File> watchedHierarchiesBuilder = ImmutableSet.builder();
+        newWatchedFiles.visitRoots(absolutePath -> watchedHierarchiesBuilder.add(new File(absolutePath)));
+        watchedHierarchies = watchedHierarchiesBuilder.build();
+
+        if (!oldWatchedHierarchies.equals(watchedHierarchies)) {
+            if (watchedHierarchies.isEmpty()) {
+                LOGGER.info("Not watching anything anymore");
+            }
+
+            List<File> hierarchiesToStopWatching = oldWatchedHierarchies.stream()
+                .filter(oldWatchedHierarchy -> !watchedHierarchies.contains(oldWatchedHierarchy))
+                .collect(Collectors.toCollection(() -> new ArrayList<>(oldWatchedHierarchies.size())));
```

Comment:
Why not use `Collectors.toList()` here? This code is not called very often and this looks like a premature optimization. Same for further down.

Reply:
Good point, this was from a time when I did expect this code to run a lot. Changed to `ImmutableList.toImmutableList()`.